.github/workflows/ci.yml

```diff
@@ -0,0 +1,94 @@
+name: Continuous integration
+
+on:
+  pull_request:
+  push:
+    branches:
+      - master
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  gather_crates:
+    runs-on: ubuntu-latest
+    outputs:
+      members: ${{ steps.cargo-metadata.outputs.members }}
+    steps:
+      - uses: actions/checkout@v3
+
+      - id: cargo-metadata
+        run: |
+          WORKSPACE_MEMBERS=$(cargo metadata --format-version=1 --no-deps | jq -c '.packages | .[] | select(.publish == null) | .name' | jq -s '.' | jq -c '.')
+          echo "members=${WORKSPACE_MEMBERS}" >> $GITHUB_OUTPUT
+
+  test:
+    name: Test ${{ matrix.crate }}
+    runs-on: ubuntu-latest
+    needs: gather_crates
+    strategy:
+      fail-fast: false
+      matrix:
+        crate: ${{ fromJSON(needs.gather_crates.outputs.members) }}
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: r7kamura/rust-problem-matchers@d58b70c4a13c4866d96436315da451d8106f8f08 #v1.3.0
+
+      - uses: Swatinem/rust-cache@359a70e43a0bb8a13953b04a90f76428b4959bb6 # v2.2.0
+        with:
+          shared-key: msrv-cache
+          save-if: false
+
+      - uses: dtolnay/rust-toolchain@stable
+
+      - uses: Swatinem/rust-cache@359a70e43a0bb8a13953b04a90f76428b4959bb6 # v2.2.0
+        with:
+          shared-key: stable-cache
+          save-if: false
```

Comment:
From what I understand you're suggesting to cache the use of `rust-problem-matchers` and `rust-cache` ? Could we consider caching crates dependencies also ? Or maybe one of those already set does this for us ?

Reply:
no no Thomas. So, `Swatinem/rust-cache` caches the cargo registry and the `target` dir see [here](https://github.com/Swatinem/rust-cache#example-usage) for more info, [`rust-problem-matchers`](https://github.com/r7kamura/rust-problem-matchers) helps showing to the user which line the problem occured:
![image](https://user-images.githubusercontent.com/1204690/223109107-4c52dc6a-32b8-4d6d-870a-4d1404fde68b.png)

Reply:
> Swatinem/rust-cache caches the cargo registry and the target dir

Ok then I'm not 100% sure to understand why we call it twice in the same job, once before installing rust stable and one after :thinking:

Reply:
ah thanks for noticing, it was a typo. We don't need the `mrsv` cache :grinning: